From 5a5665708aeb51ed3abff1c9a4de4fbc5a12fc6f Mon Sep 17 00:00:00 2001
From: Jeegar Patel <jeegar.patel@xilinx.com>
Date: Fri, 25 Sep 2020 02:11:22 -0700
Subject: [PATCH 1679/1851] drm: xlnx: zynqmp_disp: Fix OVERFLOW_BEFORE_WIDEN
 type warnings

commit ac1de8f84f3e8316090ab6c4e4eecabdf08eff32 from
https://github.com/Xilinx/linux-xlnx.git

This patch fix OVERFLOW_BEFORE_WIDEN type coverity warnings.

Potentially overflowing expression "adjusted_mode->clock * 1000" with
type "int" (32 bits, signed) is evaluated using 32-bit arithmetic, and
then used in a context that expects an expression of type "unsigned long"
(64 bits, unsigned).

To avoid overflow, cast either "adjusted_mode->clock" or "1000" to type
"unsigned long".

Signed-off-by: Jeegar Patel <jeegar.patel@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index 18d0423463f2..73825394ca68 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -2893,15 +2893,16 @@ static int zynqmp_disp_crtc_mode_set(struct drm_crtc *crtc,
 	int ret;
 
 	zynqmp_disp_clk_disable(disp->pclk, &disp->pclk_en);
-	ret = clk_set_rate(disp->pclk, adjusted_mode->clock * 1000);
+	ret = clk_set_rate(disp->pclk,
+			   (unsigned long)adjusted_mode->clock * 1000);
 	if (ret) {
 		dev_err(disp->dev, "failed to set a pixel clock\n");
 		return ret;
 	}
 
 	rate = clk_get_rate(disp->pclk);
-	diff = rate - adjusted_mode->clock * 1000;
-	if (abs(diff) > (adjusted_mode->clock * 1000) / 20) {
+	diff = rate - (unsigned long)adjusted_mode->clock * 1000;
+	if (abs(diff) > ((long)adjusted_mode->clock * 1000) / 20) {
 		dev_info(disp->dev, "request pixel rate: %d actual rate: %lu\n",
 			 adjusted_mode->clock, rate);
 	} else {
-- 
2.31.1

